### PR TITLE
[megatron] feat: add Liger TP fused cross entropy backend

### PR DESCRIPTION
**Pull request by kolehma8**
### What does this PR do?

Adds an opt-in **public Liger TP fused linear scaled cross entropy** output head to the Megatron engine, preserving Verl's existing Triton implementation as the default.

This supersedes the initial version of this PR: there is no direct call into `liger_kernel.ops.cute`, no Verl-side chunking/padding, and no new YAML chunk/tile/capacity knobs. No Liger source changes are included.

### Usage and design

```yaml
actor_rollout_ref:
  model:
    use_fused_kernels: true
    use_liger: true  # false retains the existing Verl Triton TP output head
```

```python
from liger_kernel.ops import LigerFusedLinearScaledCrossEntropyTPFunction

nll, entropy = LigerFusedLinearScaledCrossEntropyTPFunction.apply(
    hidden, weight, labels, process_group,
    temperature=temperature, ignore_index=-100, return_entropy=True,
)
```

Verl flattens hidden/target tensors, converts targets to int64 global vocabulary IDs, and returns `-nll, entropy`. It makes one public API call; Liger owns chunking, padding, and native/fallback selection. Megatron does not use `fused_kernel_options.impl_backend` for this selector.

The dependency floor is `liger-kernel>=0.8.3`. A compatible `liger-cute-kernels` wheel is optional.

**Dependency note:** `requirements.txt` intentionally specifies `liger-kernel>=0.8.3`, not `liger-kernel[lck]>=0.8.3`. Installing the requirements alone installs the public Liger API but does **not** install LCK or guarantee native execution; without LCK, Liger can select its fallback. The published `liger-kernel[lck]==0.8.3` extra installs `liger-cute-kernels==0.8.3`, whose published wheel was built with CUDA 12.9. Our B300 CUDA 13 stack required a separate compatible CUDA 13 validation build, so the extra is not enabled unconditionally. Install the LCK wheel matching the GPU/CUDA environment separately when native execution is required; all reported native experiments did so and forbade fallback.

When LCK is present on supported BF16 GPUs, Verl configures its fixed workspace from existing engine settings:

```text
capacity = max(max_token_len_per_gpu, infer_max_token_len_per_gpu) * context_parallel_size
```

Use the same limits across ranks and colocated actor/reference engines. Configuration is immutable; it is not inferred or resized from the first observed microbatch. Fixed-size microbatches are not constrained by these token-limit settings, so the limits must cover the largest packed output-head input, including Megatron alignment padding. Missing optional LCK skips setup; an installed but broken LCK dependency raises its import error.

The native output-processor hook, legacy forward, engine path, and 1F1B helper propagate the backend. Sequence-parallel gather uses scatter-only backward for Liger because Liger already reduces the hidden gradient across vocabulary shards; the existing Triton path keeps reduce-scatter. Initialization resolves tied output weights through `shared_embedding_or_output_weight()`.

There is **zero net YAML configuration change** relative to the PR base.

### Duplicate-work check

Searched open PRs for `Megatron Liger` and `fused cross entropy`. This differs from [#5130](https://github.com/verl-project/verl/pull/5130), which integrates Megatron-Core's Blackwell-specific kernel: this integrates Liger's public API on Hopper and Blackwell with temperature and entropy. [#7686](https://github.com/verl-project/verl/pull/7686) selects response-only rows in the unfused path; this PR preserves the fused path's row selection.

### Final uncached training campaign

**32 runs, all exactly 100 completed training steps**, across seeds 42-45, two backends, and four cases. Each case used eight GPUs, BF16, PP=CP=1, remove-padding, and alternating backend order. Statistics use steps 11-100 and treat the four paired seeds, not individual steps, as replicates.

GSM8K configuration: prompt cap 256, response cap 64, **4 rollouts per prompt**, generation fetch batch 8, shuffle disabled, learning rate 1e-6, entropy coefficient 0.001, KL coefficient 0.001, and fixed 8192-token workspace capacity. TP1 uses global train/PPO minibatch 128 and PPO/log-prob microbatch 32 generated sequences/GPU; TP4 uses 24 and 24, respectively. vLLM rollout TP=1, memory utilization 0.35, eager mode. Full determinism was enabled except B300 TP4; settings were identical within each backend pair.

No rollout caching or replay was used. Normalized prompt/response records differ between backends at every step in every seed. All 16 Liger runs have native-dispatch and first actor-backward markers on all eight ranks; the benchmark guard raises if Liger selects its PyTorch fallback. The guard is experiment instrumentation, not a restriction imposed by this PR's public wrapper.

#### Actor-update time per token

Positive gain means higher throughput: `Triton ms/token / Liger ms/token - 1`, averaged over seed-level pairs. The confidence interval is a Student-t interval on those four percentage gains; the paired t-test uses the four raw timing pairs. These are nominal, uncorrected comparisons, not a broad performance guarantee.

| Case | Triton ms/token | Liger ms/token | Mean gain | 95% CI | Paired t-test p |
|---|---:|---:|---:|---:|---:|
| H200 / Qwen3-1.7B / TP1 | 0.021211 | 0.020470 | +3.64% | [+0.77%, +6.51%] | 0.0253 |
| H200 / Qwen3-14B / TP4 | 0.099773 | 0.100476 | -0.69% | [-2.42%, +1.03%] | 0.2936 |
| B300 / Qwen3-4B / TP1 | 0.032649 | 0.031990 | +2.06% | [+0.71%, +3.41%] | 0.0171 |
| B300 / Qwen3-32B / TP4 | 0.251503 | 0.246207 | +2.19% | [-2.67%, +7.06%] | 0.2424 |

#### Whole-step throughput and measured input shapes

| Case | Whole-step gain | 95% CI | First actor-backward tokens/rank |
|---|---:|---:|---:|
| H200 / Qwen3-1.7B / TP1 | +0.84% | [-1.67%, +3.35%] | 4,650-4,900 |
| H200 / Qwen3-14B / TP4 | +0.10% | [-0.54%, +0.73%] | 3,288-3,864 |
| B300 / Qwen3-4B / TP1 | +0.75% | [-0.72%, +2.22%] | 4,647-4,955 |
| B300 / Qwen3-32B / TP4 | +1.11% | [-3.47%, +5.69%] | 3,504-3,712 |

Only the two TP1 actor-update comparisons exclude zero at the nominal 95% level. Neither TP4 actor-update comparison does. **None of the whole-step comparisons excludes zero.** Token counts above are the first autograd-enabled loss call on each rank per Liger run, not a trace of every microbatch. Individual sequences remain at most 320 tokens; the packed microbatch is several thousand tokens.

#### Reward observations

Qwen3-32B shows substantial reward learning under both backends. Other cases have almost entirely zero rewards: only 3/800 nonzero-reward steps for H200 TP1, 10/800 for H200 TP4, and 5/800 for B300 TP1, versus 532/800 for B300 TP4. The 64-token response cap heavily limits the smaller models; these results do **not** establish policy equivalence or comparable generalization. Nonzero entropy/KL terms exercise backward even when task rewards are zero. Longer-response H200 pilots are running with a 2048-token response cap; the tables above remain the completed 64-token-cap campaign.

The B300 Qwen3-32B plot shows one mean curve per backend with shaded **pointwise 95% Student-t confidence intervals across four seeds** (df=3), without smoothing or clipping. These are intervals for the seed mean, not simultaneous bands or paired backend-difference intervals. Negative lower bounds are an interval-estimation artifact, not negative observed rewards. [Plotted means and interval endpoints](https://github.com/kolehma8/verl/blob/d2b1b21d8ac53d68097fbda1d4c0e5bf05f88e78/b300-reward-mean-ci.csv).

**B300 / Qwen3-32B / TP4**

B300 / Qwen3-32B / TP4: mean rewards and 95% confidence intervals

### Matched-shape output-head microbenchmarks

These compare **current Verl Triton versus native Liger**, not Liger's own fallback. Eight otherwise-idle GPUs run TP1 (eight groups) or TP4 (two groups), BF16, entropy enabled, temperature 1.0, and the same 8192-token workspace as training. TP4 timing includes the sequence-parallel hidden gather and each backend's required backward gradient communication.

For each shape and phase: eight alternating-order blocks across four seeds, ten warmups and twenty timed iterations per block. Each iteration uses the maximum CUDA-event latency across all eight ranks; block medians are averaged within seed and then across seeds. Wall-clock timing gives the same conclusions. Intervals quantify within-session repeatability, not independent-machine or multi-day variability.

**TFLOP/s below is effective per-GPU output-head throughput**, using `2*M*H*(V/TP)` forward, `4*M*H*(V/TP)` backward, and `6*M*H*(V/TP)` combined, divided by measured time. It includes wrapper/launch and communication costs and excludes extra recomputation/non-GEMM arithmetic from the FLOP numerator; it is not profiler-isolated GEMM utilization. `M` is total packed tokens, not individual sequence length.

#### Representative campaign-near shapes

| GPU / model / TP / tokens | Phase | Triton ms | Liger ms | Triton TFLOP/s/GPU | Liger TFLOP/s/GPU | Throughput gain |
|---|---|---:|---:|---:|---:|---:|
| H200 / Qwen3-1.7B / TP1 / 4,800 | forward | 7.967 | 5.553 | 374.9 | 537.9 | +43.5% |
| H200 / Qwen3-1.7B / TP1 / 4,800 | backward | 25.609 | 17.093 | 233.3 | 349.5 | +49.8% |
| H200 / Qwen3-1.7B / TP1 / 4,800 | full | 33.286 | 22.936 | 269.2 | 390.7 | +45.1% |
| H200 / Qwen3-14B / TP4 / 3,648 | forward | 3.780 | 2.563 | 375.4 | 553.6 | +47.5% |
| H200 / Qwen3-14B / TP4 / 3,648 | backward | 9.409 | 9.232 | 301.6 | 307.4 | +1.9% |
| H200 / Qwen3-14B / TP4 / 3,648 | full | 13.100 | 11.793 | 324.9 | 361.0 | +11.1% |
| B300 / Qwen3-4B / TP1 / 4,800 | forward | 7.460 | 2.885 | 500.5 | 1294.3 | +158.6% |
| B300 / Qwen3-4B / TP1 / 4,800 | backward | 18.559 | 14.824 | 402.4 | 503.8 | +25.2% |
| B300 / Qwen3-4B / TP1 / 4,800 | full | 25.879 | 17.838 | 432.9 | 628.0 | +45.1% |
| B300 / Qwen3-32B / TP4 / 3,648 | forward | 3.112 | 1.128 | 456.0 | 1258.0 | +175.9% |
| B300 / Qwen3-32B / TP4 / 3,648 | backward | 5.782 | 4.193 | 490.8 | 676.9 | +37.9% |
| B300 / Qwen3-32B / TP4 / 3,648 | full | 8.737 | 5.288 | 487.2 | 805.0 | +65.2% |

<details>
<summary>Full forward+backward token sweep</summary>

| GPU | TP | Tokens | Triton ms | Liger ms | Triton TFLOP/s/GPU | Liger TFLOP/s/GPU | Gain | 95% interval |
|---|---:|---:|---:|---:|---:|---:|---:|---:|
| H200 | 1 | 4096 | 28.344 | 18.561 | 269.8 | 412.0 | +52.7% | [+49.3%, +56.2%] |
| H200 | 1 | 4800 | 33.286 | 22.936 | 269.2 | 390.7 | +45.1% | [+44.3%, +45.9%] |
| H200 | 1 | 5120 | 34.970 | 23.360 | 273.3 | 409.2 | +49.7% | [+48.5%, +50.9%] |
| H200 | 4 | 3328 | 11.769 | 11.293 | 330.0 | 343.9 | +4.2% | [+3.8%, +4.6%] |
| H200 | 4 | 3648 | 13.100 | 11.793 | 324.9 | 361.0 | +11.1% | [+10.5%, +11.7%] |
| H200 | 4 | 4096 | 14.389 | 12.076 | 332.2 | 395.8 | +19.2% | [+17.8%, +20.5%] |
| B300 | 1 | 4096 | 21.439 | 10.459 | 445.9 | 913.9 | +105.0% | [+104.0%, +106.0%] |
| B300 | 1 | 4800 | 25.879 | 17.838 | 432.9 | 628.0 | +45.1% | [+44.5%, +45.6%] |
| B300 | 1 | 5120 | 26.639 | 18.208 | 448.5 | 656.2 | +46.3% | [+45.9%, +46.8%] |
| B300 | 4 | 3328 | 7.866 | 5.091 | 493.7 | 762.8 | +54.5% | [+54.1%, +54.9%] |
| B300 | 4 | 3648 | 8.737 | 5.288 | 487.2 | 805.0 | +65.2% | [+64.8%, +65.6%] |
| B300 | 4 | 4096 | 9.483 | 5.481 | 504.0 | 872.0 | +73.0% | [+72.2%, +73.9%] |

</details>

The isolated output head is repeatably faster overall, but not every phase is faster: **H200 TP4 at 3,328 tokens has 8.0% lower Liger backward throughput**, with a 4.2% combined gain. B300 TP1 is also shape-sensitive: Liger backward is 8.053 ms at 4,096 tokens but 14.824 ms at 4,800 tokens, reducing the combined gain from 105.0% to 45.1%. No kernel-internal cause is asserted.

At the representative TP4 shape the combined absolute saving is only 1.31 ms/call on H200 and 3.45 ms/call on B300. Other model, optimizer, communication, and rollout work can dilute these savings. This supports, but does not prove without full-update profiling, dilution and variability as an explanation for the small/inconclusive end-to-end results.

### Artifacts and reproducibility

Machine-readable results are in a [separate results branch, pinned to this commit](https://github.com/kolehma8/verl/blob/387e4ddea8dfe814fe72254ffbef93659874bf7a/README.md), not added to the implementation diff.

- [Training step metrics (3,200 rows)](https://github.com/kolehma8/verl/blob/387e4ddea8dfe814fe72254ffbef93659874bf7a/step-metrics.csv), per-case summaries and audits.
- [H200 kernel table](https://github.com/kolehma8/verl/blob/387e4ddea8dfe814fe72254ffbef93659874bf7a/h200-head-summary.csv) and [B300 kernel table](https://github.com/kolehma8/verl/blob/387e4ddea8dfe814fe72254ffbef93659874bf7a/b300-head-summary.csv), including inference-only forward, training forward, backward, and combined results.
- [Benchmark harness](https://github.com/kolehma8/verl/blob/387e4ddea8dfe814fe72254ffbef93659874bf7a/benchmark_h200_campaign_heads.py), raw per-rank timing records, and [summarizer](https://github.com/kolehma8/verl/blob/387e4ddea8dfe814fe72254ffbef93659874bf7a/summarize_campaign_head_benchmarks.py).

H200 shape arguments: TP1 `--hidden-size 2048 --tokens 4096 4800 5120`; TP4 `--hidden-size 5120 --tokens 3328 3648 4096`. B300: TP1 hidden size 2560; TP4 hidden size 5120, using the same respective token grids. Global vocabulary is 151936 throughout. Launch with `torchrun --standalone --nproc-per-node=8`, `--tp 1` or `--tp 4`, and the environment's installed Verl, Megatron, and matching Liger/LCK dependencies.

The B300 environment uses the temporary CUDA 13 LCK 0.8.3 validation build. CUDA-versioned wheel publication is separate work, not part of this PR. The H200 environment uses LCK 0.8.3 with the compatible CUDA 12 stack. Native workspace/dispatch behavior matches the final integration; final changes after training only tighten optional-import error handling and repeated-configuration handling, not native computation.

Superseded cached-rollout runs, smaller-microbatch campaigns, and the stalled Qwen3.5-family 27B experiment are excluded from all final tables. The final B300 TP4 model is **Qwen3-32B**, not the earlier 27B checkpoint.

### Validation

```text
pre-commit run --all-files --show-diff-on-failure --color=never
# All 14 hooks passed; the final import-handling edit also passed a changed-file rerun.

PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 python -m pytest -q \
  tests/utils/test_linear_cross_entropy_on_cpu.py \
  tests/workers/test_megatron_liger_tp_fused_kernels_on_cpu.py \
  tests/models/test_model_forward_fused.py
# 27 passed on the Linux/H200 environment.
```

Native public-wrapper parity against an FP32 reference passed TP1 and TP4 at 137, 777, and 4096 tokens, including nonzero hidden/local-weight gradients, ignored targets, temperature 0.8, and sequence-parallel backward. The matched-shape benchmarks additionally compare both backends' outputs and gradients at all six GPU-specific shape/topology points before timing.

The local macOS environment could not collect the Megatron tests because Ray/Triton were unavailable; the reported focused-test result is from the actual Linux GPU environment.

### Submission status

- [x] Read contribution and agent guidance; checked overlapping PRs.
- [x] Updated implementation, focused tests, dependency floor, and documentation.
- [x] Finished the uncached four-seed campaign and both matched-shape kernel sweeps.
- [ ] Human submitter reviews every changed line and reproduces relevant checks before marking ready.
- [ ] Request project CI before merge.

GitHub Copilot assisted with implementation, experiments, analysis, and PR preparation. Human review and project CI are required before merge.

### CODE REVIEW

**Comment by kolehma8** [comment]
Closing for a substantial redesign: Verl should call only the public liger_kernel.ops TP frontend, rely on Liger-owned chunking/fallback behavior, and contain no direct LCK or NVSHMEM integration surface. A new PR will be opened after the implementation and validation are revised.
